### ISSUE TITLE
AVAudioPlayer sounds are very low when playing them during video chat.

### ISSUE THREAD

**Issue by astanton** [QuickBlox/quickblox-ios-sdk#483]
We have a few buttons that are on our video chat screen and we want to add sound effects when the user presses the buttons.

I've noticed that when I call these methods to play sounds outside of the video chat the volume on them sounds fine and is in line with what you would expect.

However when I call them during a video chat, the volume is significantly lower than you would expect.  I'm calling them by creating an AVAudioPlayer with the sound, then just doing `audioPlayer.prepareSound()` and `audioPlayer.play()`.

Is there some way to control the volume of the sound while in a video chat or something?

**Comment by Raikerian** [comment]
Hi,

When are you intitializing QBRTCSoundRouter?
Also, this can be helpful: https://github.com/QuickBlox/quickblox-ios-sdk/issues/393

**Comment by astanton** [comment]
@Raikerian I am initializing it in the viewDidAppear function of the video chat view controller.

I am also seeing some other issues dealing with sounds when I am trying to play more external sounds during a voice call.  Sometimes my ringtone comes out via the receiver, and sometimes it comes out via the speaker.  I want it to always come out during the speaker.  

Does the QBRTCSoundRouter setting that you set it to (speaker or receiver) also have an effect on where the AVAudioPlayer sounds come out of?

**Comment by Raikerian** [comment]
> Does the QBRTCSoundRouter setting that you set it to (speaker or receiver) also have an effect on where the AVAudioPlayer sounds come out of?

Yes, it does has effect AVAudioPlayer.

**Comment by astanton** [comment]
@Raikerian so even if I'm using a new AVAudioPlayer instance, it will have an effect on it?

If that is the case, could you let me know what the QBRTCSoundRouter changes, so that when I play my sounds, I can set this setting to speaker, then set it back to what it was prior to playing the sound?  If that makes sense...

THanks.

**Comment by shanmarkus** [comment]
Any fix on this ?

**Comment by astanton** [comment]
Not that I know of, hadn't heard back from them since the initial reply 9 days ago.  I'm not sure how to get the volume of the sound effects "normal" like when outside of the video.  If you figure it out please let me know.

**Comment by astanton** [comment]
Anyone have any suggestions on this?  I'm still having this issue and I've been searching on here and there seems to be no solution for this.

**Comment by astanton** [comment]
I've edited the sample-video-chat app to reproduce this.  It's attached it here:

[sample-videochat-webrtc-sound-issue.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/392473/sample-videochat-webrtc-sound-issue.zip)

Here are the steps to reproduce it:

- Make sure device is not set to vibrate and turn the volume all the way up.
- After logging in, click the "Ticking sound" button on the Outgoing Call View and notice how loud the ticking sound is. For some reason, if you press it a 2nd time, the sound comes out of the receiver instead of the speaker, so just press it once for the sake of hearing how loud it should be.
- Make a video call to another user and wait to connect.
- Once it is connected, you will notice there are 2 "hang up" buttons.  Click the one on the right.  This will make the ticking sound come out of the speaker.

Notice how low the volume is on the ticking sound coming out of the speaker.  I'd expect it to be just as loud as the volume it was when you hear it on the Outgoing Call View.  But it's clearly much lower.

That is the issue I am trying to resolve and looking for a solution.

**Comment by Raikerian** [comment]
State as of now is this is a confirmed bug/not available feature.
In current version of our WebRTC sdk we are using AVAudioSessionCategoryOptionDuckOthers option, which is making all other sounds but call lower. In a future version we will provide ability to change this category options to whatever you'll need using AVAudioSessionCategoryOptions (this will also add AirPlay and Bluetooth support).

**Comment by adil-appifytech** [comment]
@Raikerian - Please update your code sample because of following reasons.

1. It's not according to latest version of swift
2. categoryOptions is read only property now!

**Comment by Raikerian** [comment]
@adil-appifytech we should definitely rework swift section for our webrtc doc in a near future, but right now I don't have access to the latest swift webrtc example. Perhaps you can post code snippets that are working here for now?

**Comment by adil-appifytech** [comment]
@Raikerian - Actually I am getting error - if you could help me fix that.

<img width="1268" alt="screen shot 2017-07-18 at 7 19 06 pm" src="https://user-images.githubusercontent.com/10629648/28321932-0d4b32c8-6bee-11e7-93c7-91984f3c4250.png">

**Comment by Raikerian** [comment]
Yeah, swift 3 doesn't have binary operation anymore. Should probably be changed to something like this:
`configuration.categoryOptions = configuration.categoryOptions | .allowBluetooth`

**Comment by adil-appifytech** [comment]
@Raikerian - Not working :(

**Comment by Raikerian** [comment]
What does it say?

**Comment by adil-appifytech** [comment]
@Raikerian - Same error

**Comment by Raikerian** [comment]
My suggested code doesn't contain '|=' operand. Are you sure the error is the same.

**Comment by adil-appifytech** [comment]
@Raikerian - Please check the image. And can you show me how it's done on Android?

<img width="1647" alt="screen shot 2017-07-18 at 7 55 00 pm" src="https://user-images.githubusercontent.com/10629648/28323747-1b3a7f42-6bf3-11e7-8938-1177e4542620.png">

**Comment by Raikerian** [comment]
then just go for:
`configuration.categoryOptions = [.duckOthers, .allowBluetooth, .allowBluetoothA2DP, .allowAirPlay]` as the default enum usage in swift. categoryOptions is not readonly.

**Comment by adil-appifytech** [comment]
Now there is a new issue. It seems BluetoothA2DP and AirPlay have conflict with iOS frameworks.

<img width="1232" alt="screen shot 2017-07-20 at 12 46 49 pm" src="https://user-images.githubusercontent.com/10629648/28406352-993000fa-6d49-11e7-8e4f-ca3276f4b4f9.png">

**Comment by Raikerian** [comment]
Just add checks for the target being iOS 10 and higher or don't use these parameters at all. I suggest you read what do they mean: https://developer.apple.com/documentation/avfoundation/avaudiosessioncategoryoptions
